app/src/api/controllers/userController.ts

```diff
@@ -0,0 +1,105 @@
+import { URL } from 'url';
+
+import express from 'express';
+
+import * as userService from '../services/user';
+import * as passwordResetService from '../services/passwordReset';
+import { TokenVerificationError } from '../services/passwordReset';
+
+function createUser(req, res) {
+    const user = req.body;
+    if (req.session.user_id) {
+        res.send({ error: 'Already signed in' });
+        return;
+    }
+
+    if (user.email) {
+        if (user.email != user.confirm_email) {
+            res.send({ error: 'Email did not match confirmation.' });
+            return;
+        }
+    } else {
+        user.email = null;
+    }
+
+    userService.createUser(user).then(function (result) {
+        if (result.user_id) {
+            req.session.user_id = result.user_id;
+            res.send({ user_id: result.user_id });
+        } else {
+            req.session.user_id = undefined;
+            res.send({ error: result.error });
+        }
+    }).catch(function (err) {
+        console.error(err);
+        res.send(err);
+    });
+}
+
+function getCurrentUser(req, res) {
+    if (!req.session.user_id) {
+        res.send({ error: 'Must be logged in' });
+        return;
+    }
+
+    userService.getUserById(req.session.user_id).then(function (user) {
+        res.send(user);
+    }).catch(function (error) {
+        res.send(error);
+    });
+}
+
+function deleteCurrentUser(req, res) {
+    if (!req.session.user_id) {
+        return res.send({ error: 'Must be logged in' });
+    }
+    console.log(`Deleting user ${req.session.user_id}`);
+
+    userService.deleteUser(req.session.user_id).then(
+        () => userService.logout(req.session)
+    ).then(() => {
+        res.send({ success: true });
+    }).catch(err => {
+        res.send({ error: err });
+    });
+}
+
+async function resetPassword(req: express.Request, res: express.Response) {
+    throw new Error('adsd');
+    if(req.body == undefined || (req.body.email == undefined && req.body.token == undefined)) {
+        return res.send({ error: 'Expected an email address or password reset token in the request body' });
+    }
+
+    if(req.body.email != undefined) {
+        // first stage: send reset token to email address
+
+        // this relies on the API being on the same hostname as the frontend
+        const { origin } = new URL(req.protocol + '://' + req.headers.host);
```

Comment:
In absence of brighter ideas, this could come from an environment variable (to avoid trusting request headers).

Reply:
My thoughts too. I wanted it to be automatic but I guess updating one variable between dev and prod is not too much of a hassle. I would suggest having `WEBAPP_ORIGIN` then, set to for example `http://localhost:3000` or `https://beta.colouring.london` etc?